### PR TITLE
feat(upload-notes): add datetimes to each DocRef header

### PR DESCRIPTION
**Pull request by mikix**
Also:
- Order the DocRefs in an encounter by datetime
- Switch from freezegun to time-machine

Unrelatedly:
- Switch from deprecated cgi module to email module for mimetype parsing.

Fixes https://github.com/smart-on-fhir/cumulus-etl/issues/298

### Checklist
- [x] Consider if documentation (like in `docs/`) needs to be updated
- [x] Consider if tests should be added

### CODE REVIEW

File: cumulus_etl/fhir/fhir_utils.py [hunk 1]
```diff
@@ -104,9 +142,9 @@ async def download_reference(client: FhirClient, reference: str) -> dict | None:
 
 def _parse_content_type(content_type: str) -> (str, str):
     """Returns (mimetype, encoding)"""
-    # TODO: switch to message.Message parsing, since cgi is deprecated
-    mimetype, params = cgi.parse_header(content_type)
-    return mimetype, params.get("charset", "utf8")
+    msg = email.message.EmailMessage()
+    msg["content-type"] = content_type
+    return msg.get_content_type(), msg.get_content_charset("utf8")
```
Comment (mikix) [inline cumulus_etl/fhir/fhir_utils.py:147]: While I was here, I noticed this and figured I'd finally switch this up.

File: cumulus_etl/upload_notes/cli.py [hunk 2]
```diff
@@ -53,6 +54,15 @@ async def gather_docrefs(
         )
 
 
+def datetime_from_docref(docref: dict) -> datetime.datetime | None:
+    """Returns the date of a docref - preferring `context.period.start`, then `date`"""
```
Comment (mikix) [inline cumulus_etl/upload_notes/cli.py:58]: This is preferring clinically-relevant time over the time-of-DocRef creation. That feels right to me, but I'm open to that being flipped.

File: tests/utils.py [hunk 3]
```diff
@@ -13,21 +13,20 @@
...
+# Pass a non-UTC time to time-machine to help notice any bad timezone handling.
 # But only bother exposing the UTC version to other test code, since that's what will be most useful/common.
 _FROZEN_TIME = datetime.datetime(2021, 9, 15, 1, 23, 45, tzinfo=datetime.timezone(datetime.timedelta(hours=4)))
 FROZEN_TIME_UTC = _FROZEN_TIME.astimezone(datetime.timezone.utc)
 
 
 # Several tests involve timestamps in some form, so just pick a standard time for all tests.
-# We ignore socketserver because it checks the result of time() when evaluating timeouts.
-@freezegun.freeze_time(_FROZEN_TIME, ignore=["socketserver"])
+@time_machine.travel(_FROZEN_TIME, tick=False)
```
Comment (mikix) [inline tests/utils.py:29]: This switch fixes a crash I was seeing between pandas and freezegun (pandas only likes to use real DateTime objects, not fake ones and freezegun was generating mocked objects). Time Machine instead changes things at a lower level, and pandas seems happy.

It also fixes the issue with `socketserver` magically (despite time not ticking forward.... :shrug:)

File: cumulus_etl/upload_notes/cli.py (hunk 2 above)
Comment (dogversioning) [reply cumulus_etl/upload_notes/cli.py:58]: I think that is probably the right call.

File: cumulus_etl/upload_notes/cli.py [hunk 4]
```diff
@@ -53,6 +54,15 @@ async def gather_docrefs(
...
+    if start := fhir.parse_datetime(docref.get("context", {}).get("period", {}).get("start")):
+        return start
+    if date := fhir.parse_datetime(docref.get("date")):
+        return date
```
Comment (dogversioning) [inline cumulus_etl/upload_notes/cli.py:62]: aw yeah there's our new friends

File: tests/utils.py (hunk 3 above)
Comment (dogversioning) [reply tests/utils.py:29]: oh that's interesting - i've had to jump in hoops around that, i'll look into it.

**Review by dogversioning — APPROVED** [review]

**Pull request merged**